Improve error message: "Format Bgra8UnormSrgb can't be multisampled"

DevJac opened the pull request:
A misleading error message was stating that "Format can't be multisampled" when using an unsupported sample count. This is misleading because often these texture formats can be multisample by using a supported sample count.

Fixes #4288

DevJac committed:
Improve error message: "Format Bgra8UnormSrgb can't be multisampled"

A misleading error message was stating that "Format can't be
multisampled" when using an unsupported sample count. This is
misleading because often these texture formats can be
multisample by using a supported sample count.

cwfitzgerald commented on wgpu-core/src/pipeline.rs:
```diff
@@ -299,8 +299,8 @@ pub enum ColorStateError {
     FormatNotColor(wgt::TextureFormat),
-    #[error("Format {0:?} can't be multisampled")]
-    FormatNotMultisampled(wgt::TextureFormat),
+    #[error("Sample count {0} is not supported by format {1:?} on this device. It may be supported by your adapter through the TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES feature.")]
```
```suggestion
    #[error("Sample count {0} is not supported by format {1:?} on this device. The WebGPU spec guarentees {2:?} samples are supported on this format. With the TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES feature, your device supports {3:?}.")]
```

Adding the lists of supported sample counts will make this a lot more actionable. The error case is a slow path, so don't worry about not writing the most efficient struct.

DevJac replied:
I think we would have to instantiate a new adapter with the TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES feature in order to know what sample counts are supported by that feature. Instantiating a new adapter seems like too much for an error message.

I did improve the error message further though, take a look.

cwfitzgerald replied:
The feature is just an internal logic gate, you should be able to (where these errors are made) call `adapter.get_texture_format_features` even with the feature off.

DevJac replied:
I don't know how to get guaranteed format features? I've seen the function `guaranteed_format_features`, but it requires the device features which I don't have access to at the points in the code where these errors are created.

DevJac replied:
I did manage to produce error messages such as:


Sample count 8 is not supported by format Bgra8UnormSrgb on this device (supported sample counts are: [1, 4]). The TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES feature enables support for sample counts [1, 2, 4, 8].


and


Sample count 3 is not supported by format Bgra8UnormSrgb on this device (supported sample counts are: [1, 2, 4, 8]). The TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES feature enables support for sample counts [1, 2, 4, 8].


That last one is a little confusing though.

cwfitzgerald replied:
The features are available from the device - in this case `self.features`.

DevJac replied:
I believe this is implemented the way you wanted with the wording you wanted. Thanks for the help.

cwfitzgerald reviewed (changes requested):
Good progress! A bit of possible improvement.

DevJac committed:
TextureFormatFeatureFlags.supported_sample_counts

DevJac committed:
Include supported sample counts in InvalidSampleCount errors

DevJac commented:
I added `TextureFormatFeatureFlags.supported_sample_counts` and included the supported sample counts in the error messages. I also updated a 3rd error message which I was using as a pattern for my previous error message changes.

DevJac commented:
I included information about spec supported sample counts and device supported sample counts, as suggested. @cwfitzgerald

DevJac commented:
Squashed the most recent commits.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -2097,6 +2097,16 @@ impl TextureFormatFeatureFlags {
     }
+
+    /// A `Vec` of supported sample counts.
+    pub fn supported_sample_counts(&self) -> Vec<u32> {
+        let all_possible_sample_counts: Vec<u32> = vec![1, 2, 4, 8, 16];
+        all_possible_sample_counts
+            .iter()
+            .filter(|sc| self.sample_count_supported(**sc))
+            .cloned()
+            .collect()
+    }
```
```suggestion
    pub fn supported_sample_counts(&self) -> Vec<u32> {
        let all_possible_sample_counts: [u32; 5] = [1, 2, 4, 8, 16];
        all_possible_sample_counts
            .into_iter()
            .filter(|&sc| self.sample_count_supported(sc))
            .collect()
    }
```

Small nit here, will auto-apply

cwfitzgerald reviewed (approved):
Thanks!

cwfitzgerald committed:
Update wgpu-types/src/lib.rs
